### ISSUE TITLE
A question about the mechanism of autograd

### ISSUE THREAD

**Issue by ascust** [apache/incubator-mxnet#13660]
I am wondering how autograd works when I call backward twice.
For example, I have a network and a loss and I do:
```
with autograd.record(True):
    network_out = ....
    loss = criterion(network_out, gt)
    autograd.backwardI(loss)
with autograd.record(True):
    network_out2 = ....
    loss = criterion(network_out2, gt2)
    autograd.backwardI(loss)

optimizer.step()
```
Since I did backward twice, does the second one overrides the first one in terms of the gradients? And actually we just update the network based on the second backward? If this is true, how do I accumulate the gradients and call "step()" to update the network altogether?

**Comment by andrewfayres** [comment]
@ascust This also seems like a very good topic for discuss.mxnet.io . You might catch the interest of wider audience there.

@mxnet-label-bot add [question, Autograd]

**Comment by NRauschmayr** [comment]
The model parameters get only updated once you call optimizer.step(). The easiest would be to just feed all the data once e.g.

```
with autograd.record(True):
    network_out = model(mx.nd.array[[input1],[input2]])
    loss = criterion(network_out, mx.nd.array[[gt],[gt2]])
    autograd.backward(loss)
optimizer.step()
```

**Comment by Ishitori** [comment]
While, optimization do happen only when you call optimizer, there is one thing you may want to keep in mind, when you are doing automatic differentiation twice. It is related to how exactly gradients are stored in `.grad` array.  

By default, when you call `attach_grad()` (or when `Trainer` object does it for you) your gradients are working in `write` mode, meaning that they will be overwritten by a subsequent call of `backward()`. Here is a quick example to explain better what I mean:

```
x = mx.nd.array([1, 2, 3])
x.attach_grad()

with autograd.record():
    y = x.square()
y.backward()

with autograd.record():
    z = x.sqrt()
z.backward()

x.grad
```

The output would be:
```
[0.5        0.35355338 0.28867513]
<NDArray 3 @cpu(0)>
```
Which is the output of gradient for `sqrt()`.

If for some reason, you want to sum gradients together, you can specify different gradient mode `grad_req=add`. Take a look into this example:

```
x = mx.nd.array([1, 2, 3])
x.attach_grad(grad_req='add')

with autograd.record():
    y = x.square()
y.backward()

with autograd.record():
    z = x.sqrt()
z.backward()

x.grad
```

In that case, the output would be:
```
[2.5       4.3535533 6.2886753]
<NDArray 3 @cpu(0)>
```

Which is sum of gradients for both functions.
And as @NRauschmayr correctly said, none of these will actually change x. If I print `x` after execution of all lines above, I still get the initial values, because I haven't used optimizer:

```
[1. 2. 3.]
<NDArray 3 @cpu(0)>
```
